[RHCLOUD-17168] Integrate the marketplace token management code with the authentication endpoints

MikelAlejoBR opened the pull request:
This MR integrates the token management —catching and retrieving— code with the existing code from the authentications endpoints. That integration is tested via unit tests.

[[RHCLOUD-17168]](https://issues.redhat.com/browse/RHCLOUD-17168)

MikelAlejoBR committed:
feature: integrate marketplace token management with authentication

Integrates token caching and requesting from the marketplace with the
rest of the authentication code.

lindgrenj6 commented on dao/authentication_dao.go:
```diff
@@ -398,6 +434,49 @@ func authFromVault(secret *api.Secret) *m.Authentication {
 	}
 
+	// Request the JWT token if the authentication is a "marketplace" authentication
+	if auth.Name == "marketplace" {
```
Super small nit - can we add a log message here saying we're going to include the jwt token in the response?

MikelAlejoBR replied:
Done. Before resolving this one: I make the assumption that the authentication's name will be stored as `"marketplace"`, but maybe I'm making a too big of an assumption. Can we somehow force in [`sources-api/db/seeds/source-types.yml`](https://github.com/RedHatInsights/sources-api/blob/master/db/seeds/source_types.yml#L267-L285) which name will the auth take when it gets persisted to the db?

MikelAlejoBR replied:
@lindgrenj6 I'm assuming that the `auth.Name` will be "marketplace", but as I say, it's an assumption. Can we force a "marketplace" auth to have the name "marketplace" somehow? Thanks in advance!

lindgrenj6 commented on dao/authentication_dao_test.go:
```diff
@@ -0,0 +1,312 @@
+	flag.Bool("integration", false, "run unit or integration tests")
+	flag.Parse()
+
+	os.Exit(t.Run())
```
I _think_ we can use the parser.ParseFlags() function for this right?

MikelAlejoBR replied:
You are right. I think I forgot to update this one on the rebase :roll_eyes:

lindgrenj6 reviewed (commented):
Looks good, just a couple small things.

MikelAlejoBR committed:
refactor: use the "parser" utility instead of manually parsing

MikelAlejoBR committed:
refactor: log marketplace token inclusion in authentication

This way we can know when a marketplace was included to be sent along
with the response.

lpichler commented on dao/authentication_dao.go:
```diff
@@ -398,6 +435,51 @@ func authFromVault(secret *api.Secret) *m.Authentication {
 	}
 
+	// Request the JWT token if the authentication is a "marketplace" authentication
+	if auth.Name == "marketplace" {
```
After another look - what do think about moving branch `auth.Name == "marketplace"` to a function ?

MikelAlejoBR replied:
Initially I thought about extracting it to a function, but since `authFromVault` is already the third nested function that gets called, I thought that having to go yet into another function would make it harder to understand.

However now I'm unsure about it :S

Let me try to develop this and send another commit to see how it looks

lpichler replied:
yea that's right, 
cool!

MikelAlejoBR committed:
refactor: extract marketplace token management to a function

This way we split responsabilities, since the token management function
will have that single responsability of managing the token and modifying
the authentication struct accordingly.

MikelAlejoBR committed:
refactor: adapt tests to focus testing on the function under test

Instead of testing the whole "authFromVault" thing which wasn't the goal
of the tests, this refactor aims at testing just the marketplace bit.
